audit newest impersonated user info in the ResponseStarted, ResponseComplete audit stage

CaoShuFeng opened the pull request:
Impersonation will automatically add system:authenticated, system:serviceaccounts group to the impersonated user info. This pr use the newest impersonated user info in the second audit event. This will help users to debug rbac problems.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
[advanced audit] audit newest impersonated user info in the ResponseStarted, ResponseComplete audit stage
```
@liggitt @sttts

liggitt commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -104,6 +104,19 @@ func NewEventFromRequest(req *http.Request, level auditinternal.Level, attribs a
+// UpdateEventAttribs updates authorizer attributes according to attribs
+func UpdateEventAttribs(ev *auditinternal.Event, attribs authorizer.Attributes) {
+	if user := attribs.GetUser(); user != nil {
+		ev.User.Username = user.GetName()
```
mutating this seems wrong. I'd want both the requesting user and effective/impersonated user in my audit log

CaoShuFeng replied:
@liggitt 
Thanks for your review.
I will research it.

sttts replied:
+1

We have to extend the event type for that I guess.

/cc @timstclair

timstclair replied:
I'm not sure if I'm following what this is doing, but `ev.User` should be the "real" user, and `ev.ImpersonatedUser` should hold the the impersonated user attributes. What are we missing?

I agree with @liggitt that the audit event should be thought of as "append only". I.e. we shouldn't change information from earlier stages.

CaoShuFeng replied:
Hi, @tallclair  
I think we can extend the event type, add a new ev.EffectiveUser to events.

The ev.User in the log is the "real" user, but it is not the "effective" user. Because impersonation will change the user info after the audit filter gathers event info.

Can we add this to the [TODO list](https://github.com/kubernetes/kubernetes/issues/48561) @tallclair ?

@ericchiang What's your opinion?

tallclair replied:
I'm confused about why we need a 3rd user info on the event -- what is the difference between `EffectiveUser` and `ImpersonatedUser`?

CaoShuFeng replied:
Usually EffectiveUser is equal ImpersonatedUser.
But when user run kubectl with `--as=username` option, the difference comes.

This tables shows the difference, we can see that some groups are automatically added.
![a](https://user-images.githubusercontent.com/20855209/28196203-9cc69e44-6881-11e7-8bc3-35ab5cf5b0a9.png)

The related codes are here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go#L87
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go#L121

CaoShuFeng replied:
If we don't want to add a new field "EffectiveUser", I think we can copy some codes form impersonation.go to audit.go, and we can also add the same groups to the impersonated user.
Then the impersonated user will become effective user.

liggitt replied:
Don't copy code, I'd expect impersonated user to hold the output of the impersonation filter, what you are calling the effective user

CaoShuFeng replied:
>  I'd expect impersonated user to hold the output of the impersonation filter, what you are calling the effective user

I aggree.

CaoShuFeng replied:
I updated the code.
Please review again, thanks.

ncdc commented:
/unassign
/assign @liggitt 
fyi @mattmoyer

ericchiang commented:
Okay, so the issue is that the API server adds additional groups to the impersonated user that the audit log isn't picking up.

Rather then updating the event log later, can we move impersonation before audit log in the handler chain and have the impersonation filter (`filters.WithImpersionation`) add another context key indicating the original user?

Something like:

```go
package request

// Existing API

func UserFrom(ctx Context) (user.Info, bool) {}
func WithUser(parent Context, user user.Info) Context {}

// NewAPI. Maybe parameter names could be improved?

// WithImpersionation indicates that the user performing the request is now impersonating the
// provided user. Further calls to WithUser on the returned context will return the impersonating
// user info.
func WithImpersionation(parent Context, impersonator, impersonating user.Info) Context {}

// ImpersionationFrom returns the full impersonation information from the request. If the request
// did not use impersonation, it returns false indicating that UserFrom should be used instead.
func ImpersionationFrom(ctx Context) (impersonator, impersonating, user.Info, ok, bool) {}
```

Then have the audit log mechanism use `ImpersionationFrom`?

cc @sttts

sttts commented:
> can we move impersonation before audit log in the handler chain and have the impersonation filter

We can. But then we loose the 403 from impersonation. @soltysh was working on a catch-up audit handler to get the 403s from authentication. I guess his approach will also work with impersonation if we switch the handler chain order. @soltysh ?

CaoShuFeng commented:
Sorry to disturb you guys  again and again.
What about just use my current solution?

1. As far as I can see, this only add the automatically added groups into the impersonated user. No info will be deleted.
2. Further more, this only effect the request with impersonated user setted but impersonated groups not setted. I most cases, this change has no effect.
3. If we document this well, I thint it will not do any harm.

Thanks very much.

soltysh commented:
> We can. But then we loose the 403 from impersonation. @soltysh was working on a catch-up audit handler to get the 403s from authentication. I guess his approach will also work with impersonation if we switch the handler chain order. @soltysh ?

That shouldn't be the problem. The issue I'm trying to address in openshift (so far) is to be able to log failed authz requests, which our current implementation omits due to design decision we've made at the beginning. @crassirostris @sttts @ericchiang not sure if we want to address this at some point in k8s?

ericchiang commented:
/ok-to-test

Some possible options:

* Merge as is.
* Remove impersonation logic from audit.NewEventFromRequest and just do it here. Don't capture request impersonation headers, only the actual user that gets authenticated.
* Differentiate between request impersonation headers and actual impersonated user.

I think I'm for the first one.

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-etcd3

crassirostris commented:
@CaoShuFeng What's the status of this PR?

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -104,6 +106,22 @@ func NewEventFromRequest(req *http.Request, level auditinternal.Level, attribs a
+	if ae == nil || ae.Level.Less(audit.LevelMetadata) {
+		return
+	}
+	ae.ImpersonatedUser = &auditinternal.UserInfo{
```
@liggitt can you take a quick look?

soltysh commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:
```diff
@@ -142,6 +143,11 @@ func WithImpersonation(handler http.Handler, requestContextMapper request.Reques
 			}
 		}
 
+		if !groupsSpecified && username != user.Anonymous {
```
Now I'm puzzled. Why we need to do this, if we have [this snippet](https://github.com/kubernetes/kubernetes/blob/022919d1a4051fe39861fc5491967aea394b4787/staging/src/k8s.io/apiserver/pkg/audit/request.go#L76-L92) in place and audit happens before impersonation?

CaoShuFeng replied:
We have 4 [stages](https://github.com/kubernetes/kubernetes/blob/bc352342693e337b1ae5094af70265419a1247c5/staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go#L56).
This change will effect the other three stages rather than "StageRequestReceived".

We do this because [the snippet](https://github.com/kubernetes/kubernetes/blob/022919d1a4051fe39861fc5491967aea394b4787/staging/src/k8s.io/apiserver/pkg/audit/request.go#L76-L92) may not complete.

CaoShuFeng replied:
Reference:
https://github.com/kubernetes/kubernetes/pull/48184/files/3989d45adfccd57ffe1fdab15d3e04ab0262e152#diff-4bcea545afa7a35798d0ff9db0ad0e1aR121

sttts replied:
As far as I see the groups and UID might be different than in the old code. Can we drop the old snippet and use this instead? Having two code paths for the same things is confusing.

soltysh replied:
👍 to what Stefan said, replace the old code with the new one, entirely. I did not understand you correctly, initially.

CaoShuFeng replied:
Done.

soltysh commented:
@CaoShuFeng will you be able to update this one asap, as well?

CaoShuFeng commented:
> @CaoShuFeng will you be able to update this one asap, as well?

I have get off work today. I will update this tomorrow morning.

CaoShuFeng committed:
audit real impersonated user info

Log the newest impersonated user info in the second audit event. This
will help users to debug rbac problems.

sttts commented:
/lgtm
/approve no-issue
